ui/store/actions.js

```diff
@@ -51,9 +52,120 @@ import * as actionConstants from './actionConstants';
 
 let background = null;
 let promisifiedBackground = null;
-export function _setBackgroundConnection(backgroundConnection) {
+
+const actionRetryQueue = [];
+const promisifiedActionRetryQueue = [];
+
+// Function below is used to add action to queue
+const addToActionQueue = (queue, actionId, request, resolveHandler) => {
+  if (queue.some((act) => act.actionId === actionId)) {
+    return;
+  }
+  queue.push({
+    actionId,
+    request,
+    resolveHandler,
+  });
+};
+
+// Function below is used to remove action from queue on successful completion
+const removeFromActionQueue = (queue, actionId) => {
+  const index = queue.find((act) => act.actionId === actionId);
+  queue.splice(index, 1);
+};
+
+// Function below invokes promisifiedBackground method in MV2 content
+// In MV3 context the execution is:
+//   1. action is added to retry queue, along with resolve handler to be executed on completion of action
+//   2. is streams are connected promisifiedBackground method is called
+//   3. on completion of action either successfully or by throwing exception, action is removed from retry queue
+const submitRequestToBackground = (
+  method,
+  args = [],
+  actionId = new Date().getTime(),
+) => {
+  if (isManifestV3) {
+    return new Promise((resolve) => {
+      addToActionQueue(
+        promisifiedActionRetryQueue,
+        actionId,
+        { method, args },
+        resolve,
+      );
+      if (background.connectionStream.readable) {
+        try {
+          promisifiedBackground[method](...args).then((result) => {
```

Comment:
all errors within `new Promise(()=>{})` must be thrown synchronously or wired up into reject, so we need to be careful not to call out to functions that could have an asynchronous step.

This implementation, if `promisifiedBackground[method](...args)` gets rejected, will not catch it and cause an unhandled rejection.

Reply:
I'll dare claim it'd be beneficial to extract the queue mechanics into a separate file and/or make it testable so that we can proceed to TDD the way forward. I can help with that if someone can wire things up and create a single test that covers this to help get me started.